Practice 04

mseheda opened the pull request:
#### Summary of Changes:
- **Removed Transaction Logs:**
  - Deleted multiple transaction log files related to doctor addition.

- **Updated `DoctorsController` Class:**
  - Added new dependencies (`Logger`).
  - Enhanced error handling for doctor addition and retrieval.
  - Improved logging mechanism.

- **Added `Logger` Class:**
  - Implemented logging functionalities for information, error, and warning messages.
  
- **Modified `Doctor` Class:**
  - Minor code formatting changes.

- **Updated `Program.cs`:**
  - Added service registrations for `DoctorService` and `Logger`.
  - Configured Swagger for API documentation.

- **Refactored `DoctorService` Class:**
  - Introduced new validation and logging mechanisms.
  - Added a constant for maximum number of doctors and static property for tracking total doctors added.
  
- **Updated `IDoctorService` Interface:**
  - Modified to reflect changes in `DoctorService`.

- **Removed `Startup` Class:**
  - Configuration moved to `Program.cs`.

#### Impact:
- Improved logging and error handling in the `DoctorsController`.
- Enhanced service configuration and API documentation setup.
- Streamlined doctor management logic in `DoctorService`.
- Consolidated configuration into `Program.cs` for better maintainability.

mseheda committed:
Implemented Logger service and integrated it into Program.cs

Additionally, I rewrote Program.cs similar to a basic Web API project.

mseheda committed:
Implemented logging for adding new doctor through logger

mseheda committed:
Deleted logging into file [.log]

mseheda committed:
added MaxDoctors property

fixed GET request, now it is working properly

mseheda committed:
wip

mseheda committed:
Validation added

liudmyla-zinchenko commented on Healthcare-Hospital-Management-System/Healthcare-Hospital-Management-System/Program.cs:
```diff
@@ -1,20 +1,30 @@
+
+builder.Services.AddEndpointsApiExplorer();
+builder.Services.AddSwaggerGen();
+builder.Services.AddSingleton<IDoctorService, DoctorService>();
```
Must be AddScoped instead of AddSingleton
https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.servicecollectionserviceextensions.addscoped?view=net-8.0

liudmyla-zinchenko commented on Healthcare-Hospital-Management-System/Healthcare-Hospital-Management-System/Program.cs:
```diff
@@ -1,20 +1,30 @@
+builder.Services.AddSwaggerGen();
+builder.Services.AddSingleton<IDoctorService, DoctorService>();
+
+builder.Services.AddSingleton<Logger>();
```
Must be AddScoped instead of AddSingleton

liudmyla-zinchenko commented on Healthcare-Hospital-Management-System/Healthcare-Hospital-Management-System/Controllers/DoctorsController.cs:
```diff
@@ -1,52 +1,59 @@
-    [Route("[controller]")]
-    public class DoctorsController : ControllerBase
+    private readonly IDoctorService _doctorService;
+    private readonly Logger _logger;
```
In scope of this practice is the following point:
- Implement DI in 3 ways (constructor, property, method).
In the code you provided I see only DI using constructor but not with property and method.
Example of DI with property we had Logger at the practice, example of DI with method we had FileStream.

Please rework your PR otherwise you cannot receive the maximum for this assignment.

liudmyla-zinchenko reviewed (changes requested):
Please rework your PR according to the requirements otherwise you cannot receive the maximum for this assignment.

mseheda committed:
problems resolved

mseheda committed:
Property Injection for the Logger

mseheda commented:
I have now reworked my project due to requirements. I believe this now fully meets the assignment requirements. Please let me know if further adjustments are needed.

liudmyla-zinchenko commented on Healthcare-Hospital-Management-System/Healthcare-Hospital-Management-System/Services/DoctorService.cs:
```diff
@@ -1,66 +1,52 @@
-            byte[] messageBytes = System.Text.Encoding.UTF8.GetBytes($"DateTime: {DateTime.UtcNow}, Message: {message}\n");
-            _transactionLogFileStream.Write(messageBytes, 0, messageBytes.Length);
+            message = $"Cannot add doctor {doctor.Name}. Maximum number of doctors reached.";
+            Logger?.LogError(message);
```
You are using Logger defined at line 12, as property. 
But you do not use logger passed as parameter in the method. I see in the previous commit you used a parameter but then for some reason you changed it to the property.

By assignment requirements, you need to have both at the same time.

For now, you have unused parameter in the method - Logger logger

mseheda committed:
fix: register INotificationService in DI container

- Implemented and registered `NotificationService` as the concrete implementation of `INotificationService` in the DI container.
- Ensured the `SendNotification` method outputs data to a file (`notifications.txt`) using `StreamWriter`.

mseheda commented:
Thank you for the feedback. I have implemented **property-injected Logger** and **method-injected NotificationService** to demonstrate both injection methods as per the assignment requirements.

liudmyla-zinchenko reviewed (approved):
Good progress
